src/jitlog/generator.lua

```diff
@@ -0,0 +1,358 @@
+  element_size: Size of elements in the varible length field in bytes. fieldesize = buflen * element_size
+  implicitlen: the length of a string is implictly determined with strlen in the logger function
+]]
+
+function parser:parse_msg(def)
+  assert(def[1], "message definition contains no fields")
+  assert(def.name, "message definition has no name")
+  
+  local fieldlist = {}
+  local fieldlookup = {}
+  local vlen_fields = {}
+  --Don't try to pack into the id if we have a base
+  local offset = not def.base and 8 or 32
+  local idsize = 8 -- bits used in the message id field 24 left
+  local vcount = 0
+
+  local function add_field(f, insert_index)
+    local name = f.name
+    assert(name, "no name specifed for field")
+    if fieldlookup[name] then
+      error("Duplicate field '"..name.."' in message "..def.name)
+    end
+    fieldlookup[name] = f
+    if insert_index then
+      table.insert(fieldlist, insert_index, f)
+    else
+      table.insert(fieldlist, f)
+    end
+  end
+
+  add_field({name = "msgid", type = "u32", noarg = true, writer = "msgid"})
+
+  for i, field in ipairs(def) do
+    local name, ftype, argtype, length = parse_field(field)
+    local t = {name = name, type = ftype, argtype = argtype}
+    local typeinfo = self.types[ftype]
+
+    add_field(t)
+
+    local bitsize = 32 -- size in bits
+
+    if not typeinfo then
+      t.type = "bitfield"
+      bitsize = tonumber(ftype)
+      assert(bitsize or bitsize < 32, "invalid bitfield size")
+      t.bitsize = bitsize
+    elseif ftype == "bool" then
+      t.type = "bitfield"
+      t.bitsize = 1
+      t.bool = true
+      bitsize = 1
+    elseif typeinfo.vsize or length then
+      vcount = vcount + 1
+      vlen_fields[vcount] = t
+      t.vindex = vcount
+      t.vlen = true
+      t.ptrarg = true
+      t.buflen = length
+      t.element_size = typeinfo.element_size or typeinfo.size
+      -- Adjust the typename for primitive types to be an array
+      if not typeinfo.element_size  then
+        t.type = ftype.."[]"
+        typeinfo = self.types[t.type]
+        assert(typeinfo, "unknown buffer type")
+      end
+
+      -- Was a length field specified
+      if not length then
+        if def.use_msgsize == t.name then
+          t.buflen = "msgsize"
+        elseif ftype == "string" then
+          length = t.name.. "_length"
+          assert(not fieldlookup[length], "cannot add automatic field length because the name is already taken")
+          add_field({name = length, noarg = true, type = "uint32_t"})
+          t.buflen = length
+          t.implicitlen = true
+        else
+          error("variable length field "..t.name.."did not specify a length field")
+        end
+      end
+    else
+      bitsize = self.types[ftype].size * 8
+    end
+
+    if bitsize < 32 then
+       --Check if we can pack into the spare bits of the id field
+      if (idsize + bitsize) <= 32 then
```

Comment:
Are the brackets really needed?

Reply:
Probably not I'm just overly defensive with them to make code more readable

Reply:
I think it makes it marginally less readable (my first reaction was "ut-oh, brackets; must be some weird precedence stuff going on", which I suspect was completely incorrect). Unless there is something unusual about that expression, let's drop the brackets.

Reply:
fixed